src/wasm.rs

```diff
@@ -39,3 +48,187 @@ fn init_panic_hook() {
+    DAG_STORES.with(|dag_stores| {
+        dag_stores.borrow_mut().remove(k);
+    });
+}
+
+lazy_static! {
+    static ref TRANSACTION_COUNTER: AtomicU32 = AtomicU32::new(1);
+}
+
+thread_local! {
+    static TXNS: RefCell<HashMap<u32, Transaction<'static>>> = Default::default();
+}
+
+fn insert_transaction(txn_id: u32, txn: Transaction<'static>) -> () {
+    // todo!()
+    TXNS.with(|txns| async {
+        let mut txns = txns.borrow_mut();
+        txns.insert(txn_id, txn);
+    });
+}
+
+#[wasm_bindgen]
+pub async fn open(db_name: String, use_memstore: bool) -> Result<String, js_sys::Error> {
+    if db_name.is_empty() {
+        return Err(js_sys::Error::new("db_name must be non-empty"));
+    }
+
+    if get_dag_store(&db_name).is_some() {
+        return Err(js_sys::Error::new(&format!(
+            "Database \"{}\" has already been opened. Please close it before opening it again",
+            db_name
+        )));
+    }
+
+    let kv: Box<dyn Store> = if use_memstore {
+        Box::new(MemStore::new())
+    } else {
+        match IdbStore::new(&db_name).await {
+            Err(e) => {
+                return Err(js_sys::Error::new(&format!(
+                    "Failed to open \"{}\": {}",
+                    db_name, e
+                )))
+            }
+            Ok(store) => Box::new(store),
+        }
+    };
+
+    let lc = LogContext::new();
+    let client_id = sync::client_id::init(kv.as_ref(), lc.clone())
+        .await
+        .map_err(to_debug)
+        .map_err(|s| js_sys::Error::new(&s))?;
+
+    // insert_kv_store(db_name.clone(), kv.clone());
+
+    let store = Arc::new(dag::Store::new(kv));
+
+    do_init(store.as_ref(), lc.clone())
+        .await
+        .map_err(|e| js_sys::Error::new(&to_debug(e)))?;
+    insert_dag_store(db_name, store);
+    Ok(client_id)
+}
+
+#[wasm_bindgen]
+pub async fn open_transaction(
+    db_name: String,
+    name: Option<String>,
+    mutator_args: Option<String>,
+    basis: Option<String>,
+    original_hash: Option<String>,
+) -> Result<u32, js_sys::Error> {
+    use OpenTransactionError::*;
+
+    let rebase_opts = match (basis, original_hash) {
+        (None, None) => None,
+        (None, Some(_)) | (Some(_), None) => panic!("Invalid arguments"),
+        (Some(basis), Some(original_hash)) => Some(RebaseOpts {
+            basis,
+            original_hash,
+        }),
+    };
+
+    let lc = LogContext::new();
+
+    let store = match get_dag_store(&db_name) {
+        None => return Err(js_sys::Error::new("Database is not open")),
+        Some(store) => store,
+    };
+
+    // let x: &'static dag::Store = store.as_ref();
+
+    let txn = match name {
+        Some(mutator_name) => {
+            let mutator_args = mutator_args.ok_or(ArgsRequired)?;
+
+            let lock_timer = rlog::Timer::new();
+            debug!(lc, "Waiting for write lock...");
+            let dag_write = store.write(lc.clone()).await.map_err(DagWriteError)?;
```

Comment:
But this one complains that `store` does not live long enough. `store.clone()` does not help.

I feel pretty clueless and I do not understand how this (RC) is supposed to work?

I also tried making `write` take `self: Arc<Self>` but that took me down a rabbit hole where every single file was touched with hundreds of `clone()` calls all over. Several hours later I ended up with `DerefMut` not implemented errors and I would have to wrap a lot of code in `RwLocks` and I cried myself to sleep...